### PR TITLE
[Merged by Bors] - Refactor `mesh.go` logic into `SVM.ApplyLayer`

### LINKED ISSUE #2831: Implement `ApplyLayer` in SVM package
```go
func (s *SVM) ApplyLayer(layerID types.LayerID, transactions []types.Transaction, rewards []types.Reward)
```
Method called when a layer is applied, after Hare/Tortoise validation. This functionality is currently spread across different places and I'm not sure it's atomic (it should be).

#### Note Regarding `types.Reward`
This may not be a suitable type for this method's parameter, as it holds much more information than is needed (likely only an address and amount are required). I've used it because it was there, but we should definitely consider creating a special type for this.

### Called From
- `mesh/mesh.go` -> `Mesh.applyState`

### Existing Functions
- `TransactionProcessor.ApplyRewards`
- `TransactionProcessor.ApplyTransactions`
- `TransactionProcessor.DB.Commit`
- `TransactionProcessor.GetStateRoot`

### SVM Functions
- `IncreaseBalance`
- `Deploy`
- `Spawn`
- `Call`
- `Commit`

### PR DESCRIPTION
**Pull request by neysofu**
## Motivation

Closes #2831. It also brings forward #2069 by replacing `big.Int` with `uint64` in a few places.

## Changes

This PR completely replaces the `mesh.go:txProcessor` API with `mesh.go:svm`, which adds the `ApplyLayer` method. Some logic is then moved into the `svm` package.

## Test Plan

Relying on the current testing infrastructure as no functional changes are expected.

## DevOps Notes
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
it came up some time ago that we can't fit balance into 64 bits - https://github.com/spacemeshos/go-spacemesh/issues/2718#issuecomment-905974522 . if that's the case we should move not to uint64, but to big.Int or uint256/128 using custom lib. can we clarify that before moving forward with a proposed refactoring?

**Comment by neysofu** [comment]
@dshulyak absolutely, I was under the impression that we wanted to move forward with `uint64`, my bad! Most of this refactoring is unrelated to reward types and the work that I've done already is easily reversible. I'll put it on hold. @noamnelke I'll keep this PR focused on `SVM.ApplyLayer` and not deal with reward types.

**Comment by neysofu** [comment]
I see most of the code that I'm touching is still based on `*big.Int` (`uint64` as well, but we've established already that's not okay). I'm not sure if `*big.Int` the final decision, but I guess I should conform to that for all new code. @dshulyak what do you think?

**Comment by dshulyak** [comment]
> I'm not sure if *big.Int the final decision, but I guess I should conform to that for all new code. @dshulyak what do you think?

i think the same. based on the latest observations it is safer to assume that we won't be able to use uint64

**Comment by noamnelke** [comment]
Added this comment: https://github.com/spacemeshos/go-spacemesh/issues/2718#issuecomment-941092617

Let's use `uint64` 😃

File: activation/activationdb_test.go
Comment (countvonzero) [inline activation/activationdb_test.go]: nit: rm variable names

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: why is it named Blocks?

File: common/types/transaction.go
Comment (countvonzero) [inline common/types/transaction.go]: maybe RewardInfo, RewardRecipient?
naming a struct by its contents may not be future-proof :)

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: what is this map storing?

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: i am not sure. but it seems that data inside BySmesher and Blocks can be combined somehow to make this code more intuitive? i.e. loop over the coinbase instead of smesher+block-lookups

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: what is the diff btwn TotalReward and LayerReward

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: perhaps use *types.layerRewards here and below to save copying struct. normally i wouldn't care for small/simple struct but here there are 3 copies made for log/report/write.

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: any chance report and write can be combined? they seem to do the same loop, just slightly different output

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go]: rm or log?

File: svm/api.go
Comment (countvonzero) [inline svm/api.go]: can we avoid using named returned variables? i have seen too many subtle bugs like assignment by accident.

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go]: Removed. This slipped through, it was never intended to make upstream.

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go]: `Total` is `LayerReward` plus layer fees. I've removed `Total` and turned it into a method which calculates it on the fly.

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go]: It was supposed to indexes into `Blocks`, but it was ugly as hell and confusing. I think it's better now.

File: mesh/mesh.go
Comment (noamnelke) [inline mesh/mesh.go]: Not sure the interface should be called `svm`. We should call it something based on the role of this object in the local context - possibly `state`. `state.ApplyLayer()` makes sense, all the other functions, too. SVM is just an implementation of the state component.

File: svm/api.go
Comment (noamnelke) [inline svm/api.go]: I think this is a bad idea. The purpose of creating the `SVM` struct is to gain precise control over the methods it exposes. I think that this is a bad shortcut that will make the next phases of the SVM transition harder.

With this change, `SVM` exposes the same api as `TransactionProcessor` allowing us to use it in its place and making it harder to see what methods we still need to either implement in `SVM` or eliminate.

If there are methods that need to eventually be implemented in `SVM` - let's create them and call the matching method in `TransactionProcessor` explicitly. If there are methods that **don't** need to be there eventually, but handling them properly is unrelated to this PR - you can give them the same treatment, but add a `TODO` to remove them.

File: mesh/mesh.go
Comment (noamnelke) [reply mesh/mesh.go]: Also, most of the methods here make sense to be implemented in the `SVM`, with one exception: `ValidateAndAddTxToPool`. Does the transaction processor currently have access to the mempool? I think we should separate it, but not in this PR - for now you can implement this method - but add a `TODO` to remove it.

**Review by noamnelke — CHANGES_REQUESTED** [review]

File: svm/api.go
Comment (neysofu) [reply svm/api.go]: Ok, so to recap:

- `SVM` is not a good name for the interface. I understand your point, which makes sense. We can use `state` rather than the old `txProcessor`, I think that's okay.
- Exposing `SVM.TransactionProcessor` is simply a step towards full integration. This way, most of the code stays the same - because all previously used methods are also available on `SVM`. This was a plus in my book - incremental changes moving the abstraction layer from `txProcessor` to `SVM`, but of course the downside is that indeed you can lose track of where we are in the integration refactoring process.

Comment (neysofu) [reply svm/api.go]: I'll undo this change and implement all methods on `SVM` directly, which will then call into `TransactionProcessor`. In a future PR, we can decide which methods to keep and which ones to discard; I'll make sure to leave a `TODO` note to help us remember. Sounds good?

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go:968]: how is this possible?

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go:973]: i think this should be error. these blocks have been previously verified

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go:540]: is this log message correct? these are all valid blocks...

Comment (countvonzero) [inline mesh/mesh.go:540]: also, is this an error? is it possible that there is no coinbase for a layer?

File: mesh/mesh.go
Comment (countvonzero) [inline mesh/mesh.go:976]: this list may contain duplicate elements. is that expected by the consumers?

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go:968]: This whole portion was left untouched by the current PR. Maybe we can investigate in the future in a follow-up PR? I'm not really familiar with this code.

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go:973]: Same as above, this is not super relevant but I think it should be addressed in a separate PR.

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go:540]: Unsure; I left this untouched. I believe some tests were failing without custom handling in this `else` block (triggers division by zero). We can easily remove this if it proves not to be necessary.

Comment (neysofu) [reply mesh/mesh.go:540]: The log message is inside the `else` block, which deals with the situation where there are no (valid) blocks at all. I believe it's correct, am I missing something?

File: mesh/mesh.go
Comment (neysofu) [reply mesh/mesh.go:976]: I think it is, it definitely is for all reward calculation logic and `state` code. It's expected that a miner can submit more than one block per layer.

**Review by noamnelke — APPROVED** [review]
I think it looks good to merge. There's obviously lot's of follow up work. Will you create issues for everything and mention them in the epic?

**Comment by neysofu** [comment]
On it!

**Comment by neysofu** [comment]
bors try

**Comment by neysofu** [comment]
bors merge

**Pull request closed without merging**
